### DISCUSSION TITLE
Future of gaming and drawing in new blood

### DISCUSSION THREAD

**Discussion by darinlh** [godotengine/godot-proposals#3996] [Engine Core]
When looking at the current game market as a whole I see a few beacons of success.

Co-op / self-hosted servers and deep "MOD" support.

Starting with Bethesda style games that shipped 'Dev kits" the modding community has added innumerable hours of content and game play.

Then along comes self server hosting games such as Ark, Conan, Valhiem, Rust, etc for small communities creating multiple profitable businesses such as g-portal etc..

Now we are starting to see more interactive models such as Twitch viewers able to "spawn in" challenges for the players through APIs that tie to Twitch and discord. Even Facebook is jumping in with Meta where you can "portal" between games aka going game to game inside their "world" and Microsoft has already captured some of this with "GamePass" and of course Steam with its "Steam Deck". 

If you doubt this future ask why Unity and Unreal deny any license to allow OS shell style functioning of their engine.

The question I have is "Are there any plans to build in co-op server support, full "MOD" support/framework, and APIs to allow network interaction in Godot?"

Another subject that may need to be in a separate thread is adding in "starter templates" into libraries that can be modified so beginner devs don't just get a"blank page" to start, an example would be a basic "survival" game that actually works out of the box and beginners could dig into and modify to start learning from.

**Comment by golddotasksquestions** [comment]
> The question I have is "Are there any plans to build in co-op server support, full "MOD" support/framework, and APIs to allow network interaction in Godot?"

Server based coop and network interaction in Godot is done using the [High-level Multiplayer Network API](https://docs.godotengine.org/en/stable/tutorials/networking/high_level_multiplayer.html). Mod support is explained [here](https://docs.godotengine.org/en/stable/tutorials/export/exporting_pcks.html).

There are tons of community tutorials for these (for example on Youtube), demos in the Asset Library, and official docs. Check em out!

**Reply by darinlh** [reply]
I understand that there exists a low level API that allows networking and even "MOD" support aand there are hundreds of videos that show how to do it vs one from the devs that explain how it "should" be done correctly and a basic downloadable example that works out of the box. Just because some folks have figured out how to duct tape together code that works does NOT mean that is the correct way.

This post is on bringing in "new" devs and "modders" into the community with a jumpstart on the correct application of the current code.

Using EPIC as an example I can login as a new dev download an example module and have either of these features in place and then extend and modify them to fit my needs (hour of code). This does not exist in the Godot environment beyond the single PC 2d platformer aka 10+ year old gaming meta.

**Comment by Ophiolith** [comment]
One thing that's great about Godot is that because of its permissive license, there's literally nothing to stop a developer just shipping a copy of the editor with their game. The 'framework' is already there :) There are also 3rd-party tools like [Gridless](https://radmatt.itch.io/gridless) that can be used for a Creation Engine-style editing experience. A lot of what you're asking for is already out there!

**Reply by darinlh** [reply]
You are correct Godot is great and the license is its greatest strength BUT by pushing folks to 3rd party apps it just makes it much easy to just go to Unity or Epic because it is much 'cheaper" to build with the vast number of assets etc in the asset libraries of those companies vs buying multiple 3rd party apps they may or may not work with Godot by default.

Isn't it much more beneficial in the long run to spend a bit of time up front and create a few basic drop in building blocks that allow a new dev to get a prototype up and running as quickly as possible creating a feeling of success in the shortest time possible vs spending months deep in code just to get one basic feature working? 

There is a reason Unity, Epic, etc provide this exact model even legos provides a few basic plans out of the box.

**Reply by Xrayez** [reply]
> A lot of what you're asking for is already out there!

Well, if it takes a third-party to do all the heavy lifting to achieve this, it's less about Godot but more about other projects that use Godot or build upon Godot. Don't forget that all those features have to be maintained (regardless whether something is done officially or by third-parties). I think the author wants to have those features to be maintained officially (which likely won't happen anyways, by the way, due to Godot's development philosophy).

For example, there are a number of issues with import code that doesn't make Godot useful out-of-the-box:

- https://github.com/godotengine/godot-proposals/issues/1632

Of course, *everything* is possible with Godot, but definitely not out-of-the-box. It will definitely take independent developers or companies to do the actual work to achieve those features themselves. But again, the experience the author describes in other engines is definitely out of sight for Godot.

However, I'm interested in modding support in Godot. But according to, say https://github.com/godotengine/godot-proposals/issues/1632#issuecomment-717284002, it's difficult to see how Godot could achieve this modding support out of the box due to resistance from the core developers, more like low priority, unfortunately.
